### PR TITLE
Add a pull request chart

### PR DESCRIPTION
**Pull request by jn2657**
# Description

In this PR, I add a pull request chart.

# ScreenShot

截圖 2021-12-25 下午10 23 43

### CODE REVIEW

**Review by hellenGUAN — DISMISSED** [review]

**Review by vitachu0528 — DISMISSED** [review]

File: src/App/component/PullRequestsPage.jsx [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        console.error(e)
+      })
+  }, [])
+
+  const getPullRequestsFromGitHub = () => {
+    const githubRepo = currentProject.repositoryDTOList.find(repo => repo.type === 'github')
+    if (githubRepo !== undefined) {
+      const query = githubRepo.url.split("github.com/")[1]
+
+      // todo need reafctor with async
```
Comment (Xanonymous-GitHub) [inline src/App/component/PullRequestsPage.jsx]:
```suggestion
      // todo need refactor with async
```

File: src/App/component/PullRequestsPage.jsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  useEffect(() => {
+    Axios.get(`http://localhost:9100/pvs-api/project/1/${projectId}`,
+      { headers: { "Authorization": `${jwtToken}` } })
+      .then((response) => {
+        setCurrentProject(response.data)
+      })
+      .catch((e) => {
+        alert(e.response?.status)
+        console.error(e)
+      })
```
Comment (Xanonymous-GitHub) [inline src/App/component/PullRequestsPage.jsx]: This code is really impressive.
Because it may be the last few reviews, although this is just for school homework, not a formal open source project, I still want to provide some suggestions, hoping to do better in the future.

- Suggest extracting these API calling processes.
  For a long time, this process should be packaged independently, and in theory, the entire PVS should only be called once, rather than "call just we want", which is actually a waste of time and efficiency.
  But since we are taking things written by others and continuing to do it, I think we don’t need to refactor the whole project because of time.
  But at least the newly added pages are something we can take into account, so try to do a little better.
  In summary, the process of these API calls can be encapsulated.
  You may see there are similar structures in many places...

- Suggest giving a name for every process function. If some code does not have a function yet, suggest creating one, reduce the frequency of real logic code being placed in `useEffect`.
  Fragmented code makes it difficult to understand the meaning and purpose of execution.

File: src/App/component/PullRequestsPage.jsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  const projectId = localStorage.getItem("projectId")
+  const jwtToken = localStorage.getItem("jwtToken")
+
+  const [open, setOpen] = useState(false);
+  const handleClose = () => {
+    setOpen(false);
+  };
+  const handleToggle = () => {
+    setOpen(!open);
+  };
```
Comment (Xanonymous-GitHub) [inline src/App/component/PullRequestsPage.jsx]: Suggest given WHAT is these `open...` used for. (rename)
It may hard to understand who uses this `open` status and state setters.

```suggestion
  const [isWhatOpen, setWhatOpen] = useState(false);
  const handleWhatClose = () => {
    setWhatOpen(false);
  };
  const handleWhatToggle = () => {
    setWhatOpen(!isWhatOpen);
  };

```

File: src/App/component/PullRequestsPage.jsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    let pullRequestListDataSortedByCreatedAt = pullRequestListData
+    let pullRequestListDataSortedByClosedAt = pullRequestListData
+    let pullRequestListDataSortedByMergedAt = pullRequestListData
+
+    // Sort data by date
+    if (pullRequestListData !== undefined) {
+      [].slice.call(pullRequestListDataSortedByCreatedAt).sort((a, b) => a.createdAt - b.createdAt);
+      [].slice.call(pullRequestListDataSortedByClosedAt).sort((a, b) => a.closedAt - b.closedAt);
+      [].slice.call(pullRequestListDataSortedByMergedAt).sort((a, b) => a.mergedAt - b.mergedAt);
+    }
```
Comment (Xanonymous-GitHub) [inline src/App/component/PullRequestsPage.jsx]: 1. Suggest not let `Data` be concatenated after `List` in a variable name.

2. Suggest always use `const` to define a variable in every place, which can reduce the browser memory usage after compiling our JS code to web assembly.

3. And also suggest extract the sort process into isolated function, which may looks like:
```js
const getPRListSortedBy = (prList, key) => prList.sort((prev, curr) => prev[key] - curr[key])

const prListSortedByCreatedAt = getPRListSortedBy(prList, 'createdAt')
const prListSortedByClosedAt = getPRListSortedBy(prList, 'ClosedAt')
const prListSortedByMergedAt = getPRListSortedBy(prList, 'MergedAt')
```

File: src/App/component/PullRequestsPage.jsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    if (pullRequestListDataSortedByCreatedAt.length > 0) {
+      // Calculate the number of pull requests for each month in the selected range
+      for (let month = moment(startMonth); month < moment(endMonth).add(1, 'months'); month = month.add(1, 'months')) {
```
Comment (Xanonymous-GitHub) [inline src/App/component/PullRequestsPage.jsx]: Suggest not put any complex logic inside the for-loop condition statements.
You may need to extract these process into single variable.

File: src/App/component/PullRequestsPage.jsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        let index
```
Comment (Xanonymous-GitHub) [inline src/App/component/PullRequestsPage.jsx]: Suggest rename this variable.
It may be confused for what is a `index` here, especially inside a for-loop.

For example:
```js
for (let i = 0; i < 5; i++) {
  let j; // <== WHAT is THIS ?????? 😢 
}

**Review by hellenGUAN — DISMISSED** [review]

**Review by Xanonymous-GitHub — DISMISSED** [review]

**Review by VincentETN — APPROVED** [review]

**Review by vitachu0528 — APPROVED** [review]

**Review by vitachu0528 — APPROVED** [review]
